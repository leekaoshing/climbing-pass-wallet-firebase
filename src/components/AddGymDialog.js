import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    selectShowAddGymDialog,
    setShowAddGymDialog
} from '../reducers/dialogSlice';
import {
    addGymToEditableUser,
    selectEditableUser
} from '../reducers/userSlice';
import { selectGyms } from '../selectors/firebase';

export function AddGymDialog() {
    const dispatch = useDispatch();
    const gyms = useSelector(selectGyms);
    const user = useSelector(selectEditableUser);
    const showAddGymDialog = useSelector(selectShowAddGymDialog);

    const handleSelect = (gymId) => {
        dispatch(addGymToEditableUser(gymId));
        handleClose();
    };

    const handleClose = () => {
        dispatch(setShowAddGymDialog(false));
    }

    return (
        <Dialog onClose={handleClose} aria-labelledby="simple-dialog-title" open={showAddGymDialog}>
            <DialogTitle id="add-gym-dialog-title">Add new gym:</DialogTitle>
            <List>
                {
                    Object.keys(gyms).map(id => {
                        if (user.passes != null && !Object.keys(user.passes).includes(id)) {
                            return (
                                <ListItem key={id} button onClick={() => handleSelect(id)}>
                                    <ListItemText primary={gyms[id].name} />
                                </ListItem>
                            )
                        } else {
                            return null;
                        }
                    })
                }
            </List>
        </Dialog>
    );
}
