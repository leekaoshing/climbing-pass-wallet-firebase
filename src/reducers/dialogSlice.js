import { createSlice } from '@reduxjs/toolkit';

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState: {
    showAboutDialog: false,
    showAddGymDialog: false,
    showConfirmationDialog: false,
    showCreateUserDialog: false,
    showPasswordResetDialog: false,
    showUpdateResultDialog: false,
  },
  reducers: {
    setShowAboutDialog: (state, action) => {
      state.showAboutDialog = action.payload;
    },
    setShowAddGymDialog: (state, action) => {
      state.showAddGymDialog = action.payload;
    },
    setShowConfirmationDialog: (state, action) => {
      state.showConfirmationDialog = action.payload;
    },
    setShowCreateUserDialog: (state, action) => {
      state.showCreateUserDialog = action.payload;
    },
    setShowPasswordResetDialog: (state, action) => {
      state.showPasswordResetDialog = action.payload;
    },
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload;
    }
  }
});

export const {
  setShowAboutDialog,
  setShowAddGymDialog,
  setShowConfirmationDialog,
  setShowCreateUserDialog,
  setShowPasswordResetDialog,
  setShowUpdateResultDialog,
} = dialogSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowAboutDialog = state => {
  return state.dialog.showAboutDialog;
}

export const selectShowAddGymDialog = state => {
  return state.dialog.showAddGymDialog;
}

export const selectShowConfirmationDialog = state => {
  return state.dialog.showConfirmationDialog;
}

export const selectShowCreateUserDialog = state => {
  return state.dialog.showCreateUserDialog;
}

export const selectShowPasswordResetDialog = state => {
  return state.dialog.showPasswordResetDialog;
}

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog;
}

export default dialogSlice.reducer;
