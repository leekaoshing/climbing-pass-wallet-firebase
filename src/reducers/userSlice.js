import { createSlice } from '@reduxjs/toolkit';

export const userSlice = createSlice({
  name: 'user',
  initialState: {
    editableUser: null,
    loadingUpdateUser: false,
    updateResult: null,
  },
  reducers: {
    addGymToEditableUser: (state, action) => {
      state.editableUser.passes[action.payload] = 0
    },
    setEditableUser: (state, action) => {
      state.editableUser = action.payload;
    },
    setLoadingUpdateUser: (state, action) => {
      state.loadingUpdateUser = action.payload;
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload;
    }
  }
});

export const {
  addGymToEditableUser,
  setEditableUser,
  setLoadingUpdateUser,
  setUpdateResult
} = userSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectEditableUser = state => {
  return state.user.editableUser;
}

export const selectLoadingUpdateUser = state => {
  return state.user.loadingUpdateUser;
}

export const selectUpdateResult = state => {
  return state.user.updateResult;
}

export const getPassDifferences = state => {
  const auth = state.firebase.auth;
  const oldPasses = state.firestore.data.users[auth.uid].passes;
  const newPasses = state.user.editableUser ? state.user.editableUser.passes : {};

  const differenceInPasses = {};
  Object.keys(newPasses).forEach(key => {
    const oldPassQuantity = oldPasses[key] === undefined ? 0 : oldPasses[key];
    const newPassQuantity = newPasses[key];
    const change = newPassQuantity - oldPassQuantity;
    if (change !== 0) differenceInPasses[key] = newPassQuantity - oldPassQuantity;
  })
  return differenceInPasses;
}

export default userSlice.reducer;
