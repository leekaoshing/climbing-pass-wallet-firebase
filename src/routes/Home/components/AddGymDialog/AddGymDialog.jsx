import {
	Dialog,
	DialogContent,
	DialogTitle,
	IconButton,
	List,
	ListItem,
	ListItemText
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import AddIcon from '@material-ui/icons/Add'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import styles from './AddGymDialog.styles'
import Tooltip from '@material-ui/core/Tooltip'

const useStyles = makeStyles(styles)

function AddGymDialog({ gymsToAdd, addGymFunction }) {
	const classes = useStyles()

	const [showAddGymDialog, setShowAddGymDialog] = useState(false)

	function openAddGymDialog() {
		setShowAddGymDialog(true)
	}

	function closeAddGymDialog() {
		setShowAddGymDialog(false)
	}

	function handleSelect(gymId) {
		addGymFunction(gymId)
		closeAddGymDialog()
	}

	function isButtonDisabled() {
		return Object.keys(gymsToAdd).length === 0
	}

	return (
		<>
			<Tooltip disableFocusListener arrow enterTouchDelay={5} title="Add gym">
				<IconButton onClick={openAddGymDialog} disabled={isButtonDisabled()}>
					<AddIcon />
				</IconButton>
			</Tooltip>
			<Dialog open={showAddGymDialog} onClose={closeAddGymDialog}>
				<DialogTitle id="add-gym-dialog-title">Add gym</DialogTitle>
				<DialogContent>
					<List>
						{
							Object.keys(gymsToAdd).map(gymId => {
								return (
									<ListItem key={`add-gym-list-${gymId}`} button onClick={() => handleSelect(gymId)}>
										<ListItemText primary={gymsToAdd[gymId].name} />
									</ListItem>
								)
							})
						}
					</List>
				</DialogContent>
			</Dialog>
		</>
	)
}

AddGymDialog.propTypes = {
	gymsToAdd: PropTypes.object.isRequired,
	addGymFunction: PropTypes.func.isRequired
}

export default AddGymDialog
