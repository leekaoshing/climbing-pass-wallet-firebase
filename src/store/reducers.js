import { combineReducers } from 'redux'
import firebase from 'react-redux-firebase/lib/reducer'
import firestore from 'redux-firestore/lib/reducer'
import locationReducer from './reducers/location'
import dialogReducer from './reducers/dialog'
import userReducer from './reducers/user'

export function makeRootReducer(asyncReducers) {
  return combineReducers({
    // Add sync reducers here
    firebase,
    firestore,
    dialog: dialogReducer,
    user: userReducer,
    location: locationReducer,
    ...asyncReducers
  })
}

export function injectReducer(store, { key, reducer }) {
  store.asyncReducers[key] = reducer // eslint-disable-line no-param-reassign
  store.replaceReducer(makeRootReducer(store.asyncReducers))
}

export default makeRootReducer
