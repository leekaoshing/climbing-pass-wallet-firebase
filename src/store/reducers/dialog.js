import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  showUpdateResultDialog: false,
}

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  resetState,
  setShowUpdateResultDialog
} = dialogSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog
}

export default dialogSlice.reducer
