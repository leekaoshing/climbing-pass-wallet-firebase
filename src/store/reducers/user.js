import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  passes: null,
  updateResult: null
}

export const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    addGymToUser: (state, action) => {
      state.passes[action.payload] = 0
    },
    setPassesForGym: (state, action) => {
      const gymId = action.payload.gymId
      const numberOfPasses = action.payload.numberOfPasses
      state.passes[gymId] = numberOfPasses
    },
    setPassesForUser: (state, action) => {
      state.passes = action.payload
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  addGymToUser,
  resetState,
  setPassesForGym,
  setPassesForUser,
  setUpdateResult
} = userSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectUserPasses = state => {
  return state.user.passes
}

export const selectUpdateResult = state => {
  return state.user.updateResult
}

export default userSlice.reducer
