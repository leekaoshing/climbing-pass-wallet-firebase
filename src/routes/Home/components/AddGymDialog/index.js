import AddGymDialog from './AddGymDialog'

export default AddGymDialog
