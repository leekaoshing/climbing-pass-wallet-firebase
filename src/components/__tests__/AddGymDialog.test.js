import {
    fireEvent, render,
    screen
} from '@testing-library/react';
import { setShowAddGymDialog } from '../../reducers/userSlice';
import wrapComponent, { makeStore } from '../../testUtils/testUtils';
import { AddGymDialog } from '../AddGymDialog';

jest.disableAutomock()

describe.skip('Add Gym Dialog', () => { // TODO Material-UI Dialog doesn't render snapshots properly
    it('renders and performs select gym action correctly', () => {

        const initialState = {
            user: {
                editableUser: {
                    passes: {}
                },
            },
            dialog: {
                showAddGymDialog: true
            },
            firebase: {},
            firestore: {
                data: {
                    gyms: {
                        'FB': {
                            name: 'Fit Bloc'
                        }
                    }
                }
            }
        };

        const store = makeStore(initialState);
        const component = wrapComponent(AddGymDialog, store);

        const { asFragment } = render(component);
        expect(asFragment()).toMatchSnapshot();
        console.log(store.getState().firestore.data.gyms);

        const gymItem = screen.getByText(/Fit Bloc/i);
        fireEvent.click(gymItem);

        expect(store.getState().user.editableUser).toStrictEqual({
            passes: {
                'FB': 0
            }
        });
        expect(store.getState().user.showAddGymDialog).toBe(false);
    });
})