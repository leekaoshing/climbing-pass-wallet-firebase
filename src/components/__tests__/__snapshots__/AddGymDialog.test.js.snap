// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Gym Dialog renders and performs select gym action correctly 1`] = `<DocumentFragment />`;
